### ISSUE TITLE
I can force ban all instance with a given process definition key

### ISSUE THREAD

**Issue by Zelldon** [camunda/zeebe#14296]
**Is your feature request related to a problem? Please describe.**

Recently we ran into an issue (during a [gameday](https://confluence.camunda.com/pages/resumedraft.action?draftId=191791284&draftShareId=5a388952-aebc-4a15-93be-bfc0fa85c240&)) where we had to ban ~2k instances on all partitions. This might be quite error-prone and involves a lot of work since we need to extract somehow all instance keys before we can ban them. 

For example, this can be done via zdb or via elasticsearch (if exporters are available). See related runbook https://github.com/camunda-cloud/runbook/blob/main/zeebe/zeebe-investigate-data.md#force-banning-all-existing-instances

**Describe the solution you'd like**

Ideally, I don't need to send all the instance keys, I can just refer to the process definition key which is faulty, and ban all instances of such a model.

**Describe alternatives you've considered**

See above.

**Additional context**

It could be that this covered by the process definition deletion, but this is not clear yet. Maybe @remcowesterhoud you can clarify whether we are canceling all instances before deleting.

**Comment by abbasadel** [comment]
ZPA tirage:
- We think this is a nice idea and will make the support engineer's (medic/oncall) life easier.
- This is not a simple feature request. There might be challenges in implementing this.
- We see 2 approaches to tackle this:
- 1. We can provide a way to ban the process definition, which then means no new instances will be created either. 
- 2. We can ban the active PI of the process definition, then there may be many active and we need to figure away to handle them at one or with batching.
- We think this is an internal technical feature that ZPA wants to work on.
- We give it `Later` but we can not size it since we still need to think about a proper solution

**Comment by megglos** [comment]
ZDP-Triage:
- removing from ZDP project, as already triaged and assigned to ZPA

**Comment by Zelldon** [comment]
The need for banning appeared again, especially since we want to avoid trial users creating new instances of a malicious process model, see https://camunda.slack.com/archives/C037W9NMATG/p1702390070706609

**Comment by nicpuppa** [comment]
ZPA Triage:
- This happened in a support case as wrote in this [issue](https://github.com/camunda/zeebe/issues/15572)
- @aleksander-dytko can you have a look into this ? As for us is related to [Suspend / resume a process definition](https://github.com/camunda/product-hub/issues/289)

**Comment by aleksander-dytko** [comment]
@Zelldon, could you explain what the situations are when we need to ban multiple instances? Why is this the preferred way to handle this kind of situation?

**Comment by korthout** [comment]
@aleksander-dytko I imagine that the need is the same between banning multiple instances vs suspend/resuming multiple instances. The need here arises from the developer on-call when a cluster has a severe incident. For an example, please look at the [gameday](https://confluence.camunda.com/pages/editpage.action?useDraft=true&spaceKey=HAN&draftId=191791284&pageId=191791134&&) where this need arose.

Banning each of the instances by hand is a lot of work, and banning a set or all instances of a certain process definition is faster/easier.

Note that banning means that the instances are unusable afterward, so this is also a severe measure to use only in severe incidents. Suspending/resuming the instances would not have this impact, and would allow users to continue their process instances at some point again.

**Comment by npepinpe** [comment]
FYI @aleksander-dytko this happened to a customer again: https://camunda.slack.com/archives/C02UMKN3DTL/p1707316527352869

**Comment by Zelldon** [comment]
I still think that such things should just be priotized by the engineering teams nothing to wait for PM. This is something we need for being able to react on incidents.

**Comment by korthout** [comment]
ZPA Planning: 
- We believe we should focus rather on suspending instances of a process definition rather than banning all instances of a process definition, as this helps towards a user feature as well
- We want to come up with a concept for suspending process definition in the coming iteration

**Comment by Zelldon** [comment]
@ana-vinogradova-camunda run into it again

**Comment by korthout** [comment]
@marcosgvieira, I think we should prioritize this as a short-term way to recover clusters affected by a rogue process definition with many process instances, as banning each of these instances individually takes too much time. Banning all process instances of the process definition is a last resort only but is occasionally necessary in incidents.

It was needed again in SUPPORT-27746 where [2M process instances needed to be banned](https://camunda.slack.com/archives/C05DH1F5TAR/p1749709313621719).
EDIT: In hindsight, these instances also needed to be canceled. But banning them first would've ensured faster cancellation without unnecessarily impacting other processes.

Long-term, we need to implement support for suspending a process definition as this allows recovery of the process instances. As @ChrisKujawa shared, that can replace the ban all instances functionality. But let's focus on enabling our engineers to resolve such incidents without much effort.

EDIT: I've increased the impact of this issue, as this issue would resolve a recurring pain.

**Comment by lenaschoenburg** [comment]
This came up once again in https://jira.camunda.com/browse/SUPPORT-27380

**Comment by npepinpe** [comment]
I agree, I would like to see this for the August patch at the very least, and I think we can make that happen. July would be even better, but obviously that's only ~1 week...

**Comment by korthout** [comment]
From a technical perspective how do we want to resolve this? Looking for your input in particular @ChrisKujawa @npepinpe @lenaschoenburg, but open to anyone to chime in.

💭 I'm assuming that this operation circumvents the logstream to avoid ending up behind millions of unprocessed records.

Generally, I see several technical choices:

1. ban the process definition or the individual process instances
   - banning the definition may slow down regular execution, as it requires an additional lookup on processing (but we could choose to cache it and rebuild the cache on recovery)
   - banning the instances may slow down the operation, as it can involve a lot of rocksb writes
2. operation only on current broker or broadcasted to all brokers
   - on the current broker is simpler to implement but harder to use (requires sending the request to all brokers)
   - broadcasting the operation across all brokers may involve requiring a retry mechanism

I thus see the following options:

1. management API directly looks up all instances on all known partitions and marks them individually as banned without interacting with partitions on other brokers
2. management API directly looks up all instances on all known partitions and marks them individually as banned and sends the same request to all other known brokers
3. management API marks the process definition as banned on all known partitions without interacting with partitions on other brokers; on processing commands for a process instance, we additionally check whether the process definition is banned
4. management API marks the process definition as banned on all known partitions and sends the same request to all other known brokers; on processing commands for a process instance, we additionally check whether the process definition is banned

My preference would be option 3 with caching of the banned definition keys, as it has the least impact on the system and is easy to build, although a bit cumbersome to use (requiring requests to all brokers)

**Comment by remcowesterhoud** [comment]
I have once again a support case that would really benefit from this functionality.

> My preference would be option 3 with caching of the banned definition keys, as it has the least impact on the system and is easy to build, although a bit cumbersome to use (requiring requests to all brokers)

Idk if an in-memory cache is sufficient. It makes it very unpredictable when a definintion is banned and when it is not. I'd rather we write it in state like we do when we ban an instance. Of course we must send the request to every ~broker~ partition leader. Banning an instance currently writes an event on the log. If we introduce one for banning a definition it'll get replicated to the followers as well.

**Comment by korthout** [comment]
@remcowesterhoud I meant that the caching was extra. We anyway need persistence otherwise it wouldn't work on restarts. But we can fill a cache on restart and keep it updated when banning other definitions. Caching would avoid this having a negative performance impact

**Comment by camundaryan** [comment]
This came up on https://jira.camunda.com/browse/SUPPORT-28360

**Comment by marcosgvieira** [comment]
🔄 **Status updated automatically (Core Features)**

**From:** Ready  
**To:** Backlog  

**Reason:** The issue does not have an assignee.
